src/rust/catloop/futures/connect.rs

```diff
@@ -46,288 +47,192 @@ use ::std::{
-) -> Poll<Result<(SocketAddrV4, Rc<DuplexPipe>), Fail>> {
-    // Check if connection request was sent.
-    if let Some(handle) = DuplexPipe::poll(&self_.catmem, qt_tx)? {
-        // Retrieve operation result and check if it is what we expect.
-        let qr: demi_qresult_t = self_.catmem.borrow_mut().pack_result(handle, qt_tx)?;
-        match qr.qr_opcode {
-            // We expect a successful completion for previous push().
-            demi_opcode_t::DEMI_OPC_PUSH => {},
-            // We may get some error.
-            demi_opcode_t::DEMI_OPC_FAILED => {
-                let cause: String = format!(
-                    "failed to establish connection (qd={:?}, qt={:?}, errno={:?})",
-                    qr.qr_qd, qt_tx, qr.qr_ret
-                );
-                error!("connect_request_sent(): {:?}", &cause);
-                return Poll::Ready(Err(Fail::new(qr.qr_ret as i32, &cause)));
-            },
-            // We do not expect anything else.
-            _ => {
-                // The following statement is unreachable because we have issued a pop operation.
-                // If we successfully complete a different operation, something really bad happen in the scheduler.
-                unreachable!("unexpected operation on control duplex pipe")
-            },
+async fn get_port(
+    catmem: Rc<RefCell<CatmemLibOS>>,
+    control_duplex_pipe: Rc<DuplexPipe>,
+    yielder: &Yielder,
+) -> Result<u16, Fail> {
+    // Issue receive operation to wait for connect request ack.
+    let size: usize = mem::size_of::<u16>();
+    let qt: QToken = control_duplex_pipe.pop(Some(size))?;
+    trace!("Read port qtoken={:?}", qt);
+
+    // Get scheduler handle from the task id.
+    let mut catmem_: RefMut<CatmemLibOS> = catmem.borrow_mut();
+    let handle: TaskHandle = catmem_.schedule(qt)?;
+    drop(catmem_);
+
+    loop {
+        // Send the connection request to the server.
+        send_connection_request(catmem.clone(), control_duplex_pipe.clone(), &yielder).await?;
+
+        // Wait on the pop for MAX_ACK_RECEIVED_ATTEMPTS
+        if let Err(e) = yielder.yield_times(MAX_ACK_RECEIVED_ATTEMPTS).await {
+            return Err(e);
         }
 
-        // Issue receive operation to wait for connect request ack.
-        let size: usize = mem::size_of::<u16>();
-        let qt_rx: QToken = if let Some(qt_rx) = qt_rx {
-            qt_rx
-        } else {
-            warn!("connect_request_sent(): qt_rx is None, allocating new qtoken");
-            self_.control_duplex_pipe.pop(Some(size))?
-        };
-
-        // Transition to the next state in the connection establishment protocol.
-        self_.state = ClientState::ConnectAckReceived { qt_rx, attempt: 0 };
+        // If we received a port back from the server, then unpack it. Otherwise, send the connection request again.
+        if handle.has_completed() {
+            // Re-acquire reference to catmem libos.
+            catmem_ = catmem.borrow_mut();
```

Comment:
What if is not possible to reborrow? This may happend if some concurrent, buggy async function yields without calling drop. Re-acquire a borrow is NOT like re-acquiring a lock. The former will cause a runtime panic, whereas the latter will spin. You should consider using `try_borrow_mut()`.

Reply:
If this happens, we should panic and fix the buggy code. I don't know what situation we should let the buggy code remain and hope that it drops the reference later.

Reply:
Agreed.